### PR TITLE
feat(dm): self-wrap-only recovery for sentPartial sends

### LINKED ISSUE #4102: DM: add self-wrap-only recovery for sentPartial to avoid duplicate recipient delivery
## Summary

`ConversationBloc` currently surfaces recipient-delivered / self-wrap-failed sends as `SendStatus.sentPartial` and offers a Retry action. That Retry re-dispatches the full `ConversationMessageSent` path, which republishes recipient wraps and can duplicate delivery to recipients who already received the message.

PR #3908 intentionally accepts that tradeoff because the current send stack has no primitive to retry only the sender self-wrap.

## Problem

The current recovery path is user-visible technical debt in the DM send flow:

- 1:1: retry after `sentPartial` can send the same DM to the recipient again.
- Group DM: retry can resend to the full recipient list even if only one successful per-recipient send had a self-wrap publication failure.

## Desired Outcome

Add a self-wrap-only recovery path so `sentPartial` can repair sender cross-device sync without re-delivering to recipients.

## Scope

- Introduce a repository/service API that can publish only the sender self-addressed NIP-17 gift wrap for an already-sent rumor.
- Preserve the `sentPartial` UX signal, but make Retry use the self-wrap-only path instead of the full send path.
- For groups, scope recovery to only the affected successful recipient sends.
- Add tests that pin no duplicate recipient publish on `sentPartial` recovery.

## Context

- Landed intentionally as a follow-up to PR #3908.
- Current tradeoff is documented in `conversation_bloc.dart`.

### PR DESCRIPTION
**Pull request by realmeylisdev**
## Description

Closes #4102 — the technical-debt follow-up to PR #3908.

When a NIP-17 send delivers to the recipient but the sender's
self-addressed gift wrap fails to publish, `ConversationBloc` emits
`SendStatus.sentPartial`. Tapping Retry on the SnackBar previously
redispatched the full `ConversationMessageSent` path, which republished
the recipient gift wrap and double-delivered the message. NIP-17 self
wraps are independent of recipient wraps — only the missing self wrap
needs to land on retry, never the recipient one. This PR adds the
self-wrap-only recovery primitive plus the matching BLoC event and
SnackBar wiring so Retry-on-`sentPartial` republishes only the sender's
gift wrap, leaving recipients untouched.

**Related Issue:** Closes #4102 (follow-up to #3908)

## Type of Change

- [x] ✨ New feature (non-breaking change which adds functionality)

## Changes

**`NIP17MessageService` (`packages/dm_repository`)**
- New `publishSelfWrap({required Event rumorEvent})`: builds + publishes
  only the sender self-addressed gift wrap for a pre-existing rumor,
  sharing the underlying self-wrap helper with `sendRumor` so the two
  paths agree on null-builder, publish-failure, and exception handling.

**`DmRepository.recoverSelfWrap(rumorId)`**
- Looks up the queue row by `rumorId`, rebuilds the rumor from
  `rumor_event_json` (preserving the receiver-side dedup key), publishes
  via `publishSelfWrap`, then deletes the row on success or marks
  `self_wrap_status: failed` for the next retry sweep.
- Idempotent against an already-sent row, defends against missing rows,
  foreign-owner rows (cross-account safety), corrupted rumor JSON, and a
  missing queue DAO.

**`DmRepository.sendGroupMessage`**
- Refactored to enqueue a queue row per recipient, mirroring
  `sendMessage`, and to write per-recipient outcomes inside the same
  transaction as the local message persist. This is what lets group
  recovery scope to only the affected successful recipient sends — the
  prior implementation shared a single send path with no queue rows, so
  partial group deliveries had nowhere to look up the rumor for replay.

**`ConversationState`**
- New `PartialSend({required List<String> rumorIds})` carries the
  per-recipient rumor ids whose self-wrap failed.
- `lastPartialSend` lives alongside `lastFailedSend` with its own clear
  flag in `copyWith`. The two states drive different recovery paths
  (full resend vs self-wrap-only), so they're tracked independently.

**`ConversationBloc`**
- On `sentPartial`, populates `lastPartialSend` (not `lastFailedSend`)
  with the rumor ids of the partial-self results.
- New `ConversationSelfWrapRecoveryRequested(rumorIds)` event with a
  `sequential()` transformer (same as `ConversationMessageSent`).
- The handler loops `recoverSelfWrap` per id, aggregates outcomes, and
  reduces `lastPartialSend.rumorIds` to only the still-failing ids so a
  second Retry tap targets exactly those.
- Removes the obsolete "TODO #3909" tradeoff comment block — the
  principled fix it described is now in place.

**`ConversationView`**
- `_onSendOutcome`'s SnackBar action branches on `sendStatus`:
  `sentPartial` dispatches `ConversationSelfWrapRecoveryRequested`,
  `failed` keeps dispatching `ConversationMessageSent`.

**l10n**
- No new keys. The existing `dmSendPartialMessage` and
  `dmSendFailedRetry` already exist in all 17 locales (added by PR
  #3908).

## Test plan

Automated:
- [x] `flutter analyze lib test integration_test` — clean.
- [x] `dart format` — no changes.
- [x] `mobile/packages/dm_repository` 217/217 passing (12 new tests:
  5 `publishSelfWrap`, 5 `sendGroupMessage` queue, 7 `recoverSelfWrap`).
- [x] `mobile/test/blocs/dm` + `mobile/test/screens/inbox` 311/311
  passing (4 new self-wrap recovery scenarios + 3 new event-equality
  tests + 3 new state-shape tests + 1 new view-test for retry-dispatch).

Pinning contracts added:
- `recoverSelfWrap` never publishes a recipient wrap (verified via
  `verifyNever(messageService.sendRumor(...))` and
  `verifyNever(messageService.sendPrivateMessage(...))` after recovery).
- `ConversationSelfWrapRecoveryRequested` handler never calls
  `dmRepository.sendMessage` or `sendGroupMessage`.
- View test: tapping Retry on the partial-delivery SnackBar dispatches
  `ConversationSelfWrapRecoveryRequested` (not
  `ConversationMessageSent`).

Manual verification (please confirm during review):
- [ ] Force a self-wrap publish failure on a 1:1 send (e.g. block the
  second `publishEvent` round in a debug build) → SnackBar shows the
  partial-delivery copy → Retry resyncs the self-wrap without sending
  a second bubble to the recipient.
- [ ] Same for a group send where one of the per-recipient self-wraps
  fails → only that one rumor's self-wrap is republished on Retry.
- [ ] After a successful recovery, leave + re-enter the conversation —
  the `outgoing_dms` row for that rumor should be gone.

## Investigation notes

The plan in `tasks/lessons.md` shape, including the architectural
choice to enqueue group sends per-recipient (so partial group recovery
has a queue-row source of truth), was captured in /plan output for
#4102 prior to implementation.

### CODE REVIEW

File: mobile/packages/dm_repository/lib/src/dm_repository.dart [hunk 1]
```diff
@@ -977,12 +977,152 @@ class DmRepository {
...
+          'recovery: $e',
+          category: LogCategory.system,
+          error: e,
+          stackTrace: stackTrace,
+        );
+        // The publish succeeded; the bookkeeping failure is degraded
+        // state, not a recovery failure. The retry sweep will re-fire
+        // recoverSelfWrap, which will short-circuit via the
+        // already-sent guard above on the next pass — except we have
+        // no easy way to mark sent without re-deleting. The row
```
Comment (NotThatKindOfDrLiz) [inline mobile/packages/dm_repository/lib/src/dm_repository.dart]: This success path still leaves a real duplicate-publish hole on bookkeeping failure. If `deleteById()` throws, the row remains `recipient: sent / self: failed`, so the next recovery pass republishes the self-wrap even though this publish already landed. `markSelfWrapStatus(id: rumorId, status: OutgoingWrapStatus.sent, eventId: result.messageEventId)` is already available and would close the loop here as a best-effort fallback.

File: mobile/packages/dm_repository/lib/src/dm_repository.dart [hunk 2]
```diff
@@ -1059,9 +1239,64 @@ class DmRepository {
           ownerPubkey: _userPubkey,
           dmProtocol: existingGroup?.dmProtocol,
         );
+
+        if (outgoingDao != null) {
+          for (var i = 0; i < rumors.length; i++) {
+            final rumor = rumors[i];
+            final result = results[i];
+            if (!result.success) continue;
+            if (result.selfWrapPublished == true) {
```
Comment (NotThatKindOfDrLiz) [inline mobile/packages/dm_repository/lib/src/dm_repository.dart]: This is now the second copy of the same recipient-success / self-wrap-failed queue transition that already exists in `sendMessage()`. Since this PR introduces the new twin, please extract the shared transition into a private helper on `DmRepository` rather than locking in two copies of the same bookkeeping logic.

File: mobile/packages/dm_repository/lib/src/dm_repository.dart [hunk 3]
```diff
@@ -977,12 +977,152 @@ class DmRepository {
...
+      throw ArgumentError.value(
+        rumorId,
+        'rumorId',
+        'queue row belongs to a different account',
+      );
+    }
+
+    // Idempotent re-tap: a concurrent recovery already landed. Return
+    // success so the caller's aggregation treats this rumor as done.
+    if (row.selfWrapStatus == OutgoingWrapStatus.sent) {
```
Comment (NotThatKindOfDrLiz) [inline mobile/packages/dm_repository/lib/src/dm_repository.dart:1006]: The idempotent guard is good defensively, but right now this PR never produces `selfWrapStatus == sent` in production. As written, this branch is only reachable from test fixtures or some future sweep path. Either make item 1 above mark `selfWrapStatus` to `sent` on delete failure, or tighten the comment so it says explicitly that this is future-facing rather than a state current code can create.

File: mobile/packages/dm_repository/lib/src/dm_repository.dart (hunk 1 above)
Comment (realmeylisdev) [reply mobile/packages/dm_repository/lib/src/dm_repository.dart]: Fixed in 73db4e4d6 — the success path now falls back to `markSelfWrapStatus(id: rumorId, status: sent, eventId: result.messageEventId)` when `deleteById` throws, so the next recovery sweep short-circuits via the idempotent already-sent guard instead of republishing. Added two tests covering the fallback path (delete-throws → marks sent + returns success) and the doubly-degraded path (both writes throw → still returns success, safe via NIP-17 receive-side dedup, surfaced via logs).

File: mobile/packages/dm_repository/lib/src/dm_repository.dart (hunk 2 above)
Comment (realmeylisdev) [reply mobile/packages/dm_repository/lib/src/dm_repository.dart]: Extracted in 73db4e4d6 — both call sites now go through `_finalizeAfterRecipientSuccess` (private async helper on `DmRepository` placed between `recoverSelfWrap` and `sendGroupMessage`). Helper runs inside the parent `runInTransaction`, so atomicity is preserved. Existing queue tests assert at the DAO-call surface and stay green unchanged.

File: mobile/packages/dm_repository/lib/src/dm_repository.dart (hunk 3 above)
Comment (realmeylisdev) [reply mobile/packages/dm_repository/lib/src/dm_repository.dart:1006]: Tightened in 73db4e4d6 — went with option 1 (mark sent on delete failure), so the guard is now reachable from production code. Updated the comment to describe both reachable paths: concurrent recovery for the same rumor, or the previous sweep's deleteById threw and the fallback marked self_wrap_status=sent.

**Review by NotThatKindOfDrLiz — CHANGES_REQUESTED** [review]
Requesting changes on three adjacent issues before merge:

1. `recoverSelfWrap()` still has a real duplicate-publish bookkeeping hole when `deleteById()` fails after a successful self-wrap publish. Please close that loop with a best-effort `markSelfWrapStatus(...sent...)` fallback so the row does not stay retryable in a state that republishes the self-wrap.
2. The `selfWrapStatus == sent` idempotent guard is not produced by this PR today, so either make it real via item 1 or tighten the comment to say explicitly that it is future-facing/defensive.
3. The recipient-success / self-wrap-failed queue transition is now duplicated between `sendMessage()` and `sendGroupMessage()`. Since this PR adds the new twin, please extract the shared bookkeeping into a private helper instead of locking in two copies.

Architecturally the recovery direction is correct, and I agree with the core fix for #4102. I am not blocking on the snackbar accessibility point because `conversation_view.dart` already explicitly announces the async snackbar outcome on `main`; the merge blockers here are the three code issues above.

Before merge, the follow-up tracking should also exist for:
- auto-sweeping `recipient: sent / self: failed` rows
- surfacing swallowed DAO bookkeeping failures instead of only logging them

**Comment by realmeylisdev** [comment]
@NotThatKindOfDrLiz — addressed all three blockers in 73db4e4d6 and filed the two follow-up tracking issues you flagged before merge. Ready for re-review.

**Code blockers — all in 73db4e4d6:**

1. **Duplicate-publish hole on `deleteById` failure.** Success path now falls back to `markSelfWrapStatus(id: rumorId, status: sent, eventId: result.messageEventId)` so the next sweep short-circuits via the idempotent guard. Doubly-degraded path (both writes throw) is logged and safe via NIP-17 receive-side dedup. New tests cover both branches.
2. **Comment on the `selfWrapStatus == sent` guard.** Went with option 1 — the guard is now reachable from production code via the fallback, and the comment describes both reachable paths (concurrent recovery; previous sweep's `deleteById` threw and fallback marked sent).
3. **Duplicated recipient-success / self-wrap-failed bookkeeping.** Extracted to private `_finalizeAfterRecipientSuccess({outgoingDao, rumorId, result})` on `DmRepository`; called inside the parent transaction from both `sendMessage` and `sendGroupMessage`. Existing queue tests assert at the DAO-call surface and stay green.

**Follow-up tracking issues:**

- #4124 — auto-sweep `recipient: sent / self: failed` rows so dormant queue rows clear without user action.
- #4127 — surface swallowed DAO bookkeeping failures in `recoverSelfWrap` / `_finalizeAfterRecipientSuccess` (today only `Log.error`, not `Reportable`).

CI green on 73db4e4d6.

**Review by NotThatKindOfDrLiz — APPROVED** [review]
Re-reviewed on 7b2ea8441ad8d26d3f78739b5c509f8b0462a874. The remaining no-tech-debt issues are addressed: recipient-failure bookkeeping is centralized alongside the success helper, and the temporary sendRumor test shim was removed rather than left behind as permanent scaffolding. I reran the relevant local verification and all GitHub checks are green on this head.
